### PR TITLE
Fixes a bug in the Upstream Pulp Replicate API

### PR DESCRIPTION
**Pull request by dkliban**
This bug was only present when domains were enabled.

fixes: #4119

### CODE REVIEW

**Review by ipanova — APPROVED** [review]

**Comment by ipanova** [comment]
any versions we want to backport this to?

**Comment by dkliban** [comment]
we should backport to 3.23, 3.28, 3.29

**Comment by dkliban** [comment]
I added a test and that required fixing another bug as part of the same PR.

File: pulpcore/tests/functional/__init__.py [hunk 1]
```diff
@@ -1063,3 +1069,37 @@ def _wget_recursive_download_on_host(url, destination):
         )
 
     return _wget_recursive_download_on_host
+
+
+@pytest.fixture()
+def non_default_domain(domains_api_client, pulp_settings, gen_object_with_cleanup):
```
Comment (mdellweg) [inline pulpcore/tests/functional/__init__.py:1075]: Why don't we make this a domain_factory right away?

File: pulpcore/tests/functional/api/test_replication.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        "api_root": pulp_settings.API_ROOT,
+        "domain": "default",
+        "username": domains_api_client.api_client.configuration.username,
+        "password": domains_api_client.api_client.configuration.password,
+    }
+    upstream_pulp = gen_object_with_cleanup(
+        upstream_pulp_api_client, upstream_pulp_body, pulp_domain=non_default_domain.name
+    )
+    # Run the replicate task and assert that all tasks successfully complete.
+    response = upstream_pulp_api_client.replicate(upstream_pulp.pulp_href)
```
Comment (mdellweg) [inline pulpcore/tests/functional/api/test_replication.py:30]: So this is really the first time we sync from one pulp into itself, right?

**Review by mdellweg — CHANGES_REQUESTED** [review]

File: pulpcore/tests/functional/__init__.py (hunk 1 above)
Comment (dkliban) [reply pulpcore/tests/functional/__init__.py:1075]: i was going to convert it on my next PR, but I'll do it now.

File: pulpcore/tests/functional/api/test_replication.py (hunk 2 above)
Comment (dkliban) [reply pulpcore/tests/functional/api/test_replication.py:30]: yes, i believe so.

**Review by mdellweg — APPROVED** [review]

**Pull request merged**